File: SDUI/src/sdui/parser.py
"""Bounded recursive descent for grammar/sdui-0.2.ebnf."""
from .ast import (Argument, Connection, Definition, Document, LayoutRule, Literal,
                  ModuleRef, Node, Reference, Row, SduiError, Span)
from .lexer import lex

RESERVED = {'sdui', 'ref', 'true', 'false', 'null', 'setHandle'}


class Parser:
    def __init__(self, source):
        self.tokens, self.source = lex(source), source.encode('utf-8')
        self.index, self.nodes = 0, 0

    @property
    def token(self):
        return self.tokens[self.index]

    def ahead(self, kind):
        return self.tokens[min(self.index + 1, len(self.tokens) - 1)].kind == kind

    def take(self, kind, value=None):
        token = self.token
        if token.kind != kind or (value is not None and token.value != value):
            self.fail(f'Expected {value or kind}, found {token.value!r}')
        self.index += 1
        return token

    def accept(self, kind):
        return self.take(kind) if self.token.kind == kind else None

    def fail(self, message, code='syntax'):
        raise SduiError(code, message, self.token.span)

    def identifier(self):
        if self.token.value in RESERVED:
            self.fail('Reserved word cannot be used as an identifier')
        token = self.take('ID')
        if '-' in token.value:
            raise SduiError('syntax', 'Hyphens are only allowed in property names', token.span)
        return token

    def span(self, start):
        return Span(start.start, self.tokens[self.index - 1].span.end, start.line, start.column)

    def parse(self):
        start = self.take('ID', 'sdui').span
        version = self.take('NUMBER')
        if self.source[version.span.start:version.span.end] != b'0.2':
            raise SduiError('version', 'Only exact profile sdui 0.2 is supported', version.span)
        self.take(';')
        refs, definitions, connections = [], [], []
        while self.token.value == 'ref':
            begin = self.take('ID').span
            self.take(':')
            alias = self.identifier().value
            path = self.take('STRING').value
            self.take(';')
            refs.append(ModuleRef(alias, path, self.span(begin)))
        while self.token.kind == 'ID' and self.ahead('='):
            begin = self.token.span
            name = self.identifier().value
            self.take('=')
            root = self.node(1)
            self.take(';')
            definitions.append(Definition(name, root, self.span(begin)))
        if not definitions:
            self.fail('At least one UI definition is required')
        while self.token.kind != 'EOF':
            begin = self.token.span
            module = self.identifier().value
            self.take('.')
            obj = self.identifier().value
            self.take('.')
            self.take('ID', 'setHandle')
            self.take('(')
            definition = self.identifier().value
            path = []
            while self.accept('.'):
                path.append(self.identifier().value)
            if not path:
                self.fail('setHandle needs a component instance path')
            self.take(')')
            self.take(';')
            connections.append(Connection(module, obj, definition, tuple(path), self.span(begin)))
        end = self.take('EOF').span.end
        return Document('sdui/0.2', tuple(refs), tuple(definitions), tuple(connections),
                        Span(start.start, end, start.line, start.column))

    def node(self, depth, parent=None):
        if depth > 64:
            self.fail('Component nesting exceeds 64', 'depth-limit')
        self.nodes += 1
        if self.nodes > 2048:
            self.fail('More than 2048 components', 'node-limit')
        start, name, role = self.token.span, None, None
        if self.token.kind == 'ID' and self.ahead('='):
            name = self.identifier().value
            self.take('=')
            if parent == 'frame' and name in {'header', 'body', 'footer'}:
                role, name = name, None
        rows, args, text, target, widget, variant = (), (), None, None, None, None
        if self.token.kind in {'[', '<'}:
            opening = self.token.kind
            self.take(opening)
            kind, close = ('frame', ']') if opening == '[' else ('group', '>')
            rows = self.rows(close, depth, kind)
            self.take(close)
        elif self.token.kind == 'STRING':
            kind, text = 'markdown', self.literal()
        elif self.token.kind == 'ID':
            target = self.identifier().value
            if self.accept('('):
                kind, widget, target = 'widget', target, None
                args = self.arguments()
                self.take(')')
            else:
                kind = 'use'
        else:
            self.fail('Expected frame, group, Markdown, widget call or component reference')
        if self.accept('*'):
            variant = self.identifier().value
        layout = self.layout() if self.token.kind == '{' else ()
        return Node(kind, name, role, rows, widget, args, text, target, variant, layout, self.span(start))

    def rows(self, close, depth, parent):
        rows = []
        if self.token.kind == close:
            return ()
        while True:
            begin = self.token.span
            items = [self.node(depth + 1, parent)]
            while self.accept(','):
                items.append(self.node(depth + 1, parent))
            rows.append(Row(tuple(items), self.span(begin)))
            if not self.accept(';') or self.token.kind == close:
                break
        return tuple(rows)

    def arguments(self):
        args = []
        if self.token.kind != ')':
            while True:
                start, name = self.token.span, None
                if self.token.kind == 'ID' and self.ahead('='):
                    name = self.identifier().value
                    self.take('=')
                args.append(Argument(name, self.value(), self.span(start)))
                if len(args) > 32:
                    self.fail('Widget has more than 32 arguments', 'argument-limit')
                if not self.accept(','):
                    break
        return tuple(args)

    def value(self):
        if self.token.kind == 'ID' and self.token.value not in {'true', 'false', 'null'}:
            start = self.token.span
            module = self.identifier().value
            self.take('.')
            obj = self.identifier().value
            self.take('.')
            self.take('@')
            member = self.identifier().value
            return Reference(module, obj, member, self.span(start))
        return self.literal()

    def literal(self):
        token = self.token
        if token.kind in {'STRING', 'NUMBER'}:
            self.take(token.kind)
            return Literal('string' if token.kind == 'STRING' else 'number', token.value, token.span)
        if token.kind == 'ID' and token.value in {'true', 'false', 'null'}:
            self.take('ID')
            return Literal('null' if token.value == 'null' else 'boolean',
                           {'true': True, 'false': False, 'null': None}[token.value], token.span)
        self.fail('Expected literal')

    def layout_value(self):
        start = self.token.span
        if self.accept('('):
            values = [self.take('NUMBER').value]
            while self.accept(','):
                values.append(self.take('NUMBER').value)
                if len(values) > 4:
                    self.fail('Padding tuple has at most four values')
            self.take(')')
            return Literal('tuple', tuple(values), self.span(start))
        if self.token.kind == 'ID' and self.token.value not in {'true', 'false', 'null'}:
            return Literal('enum', self.take('ID').value, start)
        value = self.literal()
        if value.kind == 'number' and self.token.value == 'fr':
            self.take('ID')
            return Literal('fr', value.value, self.span(start))
        return value

    def layout(self):
        self.take('{')
        rules = []
        if self.token.kind != '}':
            while True:
                start = self.token.span
                if self.token.kind == 'ID' and self.ahead('='):
                    name = self.take('ID').value
                    self.take('=')
                    value = self.layout_value()
                elif self.token.kind == 'NUMBER' and self.ahead(':'):
                    x = self.take('NUMBER').value
                    self.take(':')
                    y = self.take('NUMBER').value
                    name, value = 'ratio', Literal('ratio', (x, y), self.span(start))
                else:
                    if self.token.kind in {',', '}', 'EOF'}:
                        self.fail('Expected formatting rule')
                    while self.token.kind not in {',', '}', 'EOF'}:
                        self.index += 1
                    name = 'arrow'
                    raw = self.source[start.start:self.tokens[self.index-1].span.end].decode()
                    value = Literal('arrow', ''.join(raw.split()), self.span(start))
                spelling = self.source[start.start:self.tokens[self.index-1].span.end].decode()
                rules.append(LayoutRule(name, value, spelling, self.span(start)))
                if len(rules) > 32:
                    self.fail('More than 32 formatting rules', 'layout-limit')
                if not self.accept(','):
                    break
        self.take('}')
        return tuple(rules)


def parse(source: str) -> Document:
    return Parser(source).parse()

File: SDUI/go/parser/parser.go
package parser

import "strings"

type reader struct {
	tokens       []token
	source       string
	index, nodes int
}

func (p *reader) t() token { return p.tokens[p.index] }
func (p *reader) ahead(k string) bool {
	return p.index+1 < len(p.tokens) && p.tokens[p.index+1].kind == k
}
func (p *reader) take(k string, value ...string) token {
	t := p.t()
	if t.kind != k || len(value) > 0 && t.value != value[0] {
		fail("syntax", "Expected "+k, t.span)
	}
	p.index++
	return t
}
func (p *reader) accept(k string) bool {
	if p.t().kind == k {
		p.take(k)
		return true
	}
	return false
}
func (p *reader) ident() token {
	t := p.take("ID")
	s := t.value.(string)
	if strings.Contains(s, "-") || s == "sdui" || s == "ref" || s == "true" || s == "false" || s == "null" || s == "setHandle" {
		fail("syntax", "Reserved or invalid identifier", t.span)
	}
	return t
}
func (p *reader) span(s Span) Span { s.End = p.tokens[p.index-1].span.End; return s }

// Parse constructs a source-positioned AST, without resolving names or opening ref sources.
func Parse(source string) (doc *Document, err error) {
	defer recoverDiagnostic(&err)
	p := reader{tokens: lex(source), source: source}
	return p.document(), nil
}
func (p *reader) document() *Document {
	start := p.take("ID", "sdui").span
	v := p.take("NUMBER")
	if p.source[v.span.Start:v.span.End] != "0.2" {
		fail("version", "Only exact sdui 0.2 is supported", v.span)
	}
	p.take(";")
	d := &Document{Profile: "sdui/0.2"}
	for p.t().value == "ref" {
		s := p.take("ID").span
		p.take(":")
		a := p.ident().value.(string)
		path := p.take("STRING").value.(string)
		p.take(";")
		d.References = append(d.References, ModuleRef{a, path, p.span(s)})
	}
	for p.t().kind == "ID" && p.ahead("=") {
		s := p.t().span
		n := p.ident().value.(string)
		p.take("=")
		root := p.node(1, "")
		p.take(";")
		d.Definitions = append(d.Definitions, Definition{n, root, p.span(s)})
	}
	if len(d.Definitions) == 0 {
		fail("syntax", "At least one definition is required", p.t().span)
	}
	for p.t().kind != "EOF" {
		s := p.t().span
		m := p.ident().value.(string)
		p.take(".")
		o := p.ident().value.(string)
		p.take(".")
		p.take("ID", "setHandle")
		p.take("(")
		def := p.ident().value.(string)
		path := []string{}
		for p.accept(".") {
			path = append(path, p.ident().value.(string))
		}
		if len(path) == 0 {
			fail("syntax", "Expected component path", p.t().span)
		}
		p.take(")")
		p.take(";")
		d.Connections = append(d.Connections, Connection{m, o, def, path, p.span(s)})
	}
	end := p.take("EOF")
	start.End = end.span.End
	d.Span = start
	return d
}
func (p *reader) node(depth int, parent string) *Node {
	if depth > 64 {
		fail("depth-limit", "Component depth exceeds 64", p.t().span)
	}
	p.nodes++
	if p.nodes > 2048 {
		fail("node-limit", "More than 2048 components", p.t().span)
	}
	n := &Node{Span: p.t().span}
	if p.t().kind == "ID" && p.ahead("=") {
		name := p.ident().value.(string)
		p.take("=")
		if parent == "frame" && (name == "header" || name == "body" || name == "footer") {
			n.Role = str(name)
		} else {
			n.Name = str(name)
		}
	}
	switch p.t().kind {
	case "[", "<":
		open := p.t().kind
		p.take(open)
		close := ">"
		n.Kind = "group"
		if open == "[" {
			close = "]"
			n.Kind = "frame"
		}
		n.Rows = p.rows(close, depth, n.Kind)
		p.take(close)
	case "STRING":
		n.Kind = "markdown"
		l := p.literal()
		n.Text = &l
	case "ID":
		name := p.ident().value.(string)
		if p.accept("(") {
			n.Kind = "widget"
			n.Widget = str(name)
			n.Arguments = p.arguments()
			p.take(")")
		} else {
			n.Kind = "use"
			n.Target = str(name)
		}
	default:
		fail("syntax", "Expected UI component", p.t().span)
	}
	if p.accept("*") {
		n.Variant = str(p.ident().value.(string))
	}
	if p.t().kind == "{" {
		n.Layout = p.layout()
	}
	n.Span = p.span(n.Span)
	return n
}
func (p *reader) rows(close string, depth int, parent string) []Row {
	rows := []Row{}
	if p.t().kind == close {
		return rows
	}
	for {
		s := p.t().span
		items := []*Node{p.node(depth+1, parent)}
		for p.accept(",") {
			items = append(items, p.node(depth+1, parent))
		}
		rows = append(rows, Row{items, p.span(s)})
		if !p.accept(";") || p.t().kind == close {
			break
		}
	}
	return rows
}
func (p *reader) arguments() []Argument {
	args := []Argument{}
	if p.t().kind == ")" {
		return args
	}
	for {
		a := Argument{Span: p.t().span}
		if p.t().kind == "ID" && p.ahead("=") {
			a.Name = str(p.ident().value.(string))
			p.take("=")
		}
		a.Value = p.value()
		a.Span = p.span(a.Span)
		args = append(args, a)
		if len(args) > 32 {
			fail("argument-limit", "More than 32 arguments", p.t().span)
		}
		if !p.accept(",") {
			return args
		}
	}
}
func (p *reader) value() any {
	if p.t().kind == "ID" && p.t().value != "true" && p.t().value != "false" && p.t().value != "null" {
		s := p.t().span
		m := p.ident().value.(string)
		p.take(".")
		o := p.ident().value.(string)
		p.take(".")
		p.take("@")
		member := p.ident().value.(string)
		return Reference{m, o, member, p.span(s)}
	}
	return p.literal()
}
func (p *reader) literal() Literal {
	t := p.t()
	switch t.kind {
	case "STRING":
		p.take("STRING")
		return Literal{"string", t.value, t.span}
	case "NUMBER":
		p.take("NUMBER")
		return Literal{"number", t.value, t.span}
	case "ID":
		if t.value == "true" || t.value == "false" {
			p.take("ID")
			return Literal{"boolean", t.value == "true", t.span}
		}
		if t.value == "null" {
			p.take("ID")
			return Literal{"null", nil, t.span}
		}
	}
	fail("syntax", "Expected literal", t.span)
	return Literal{}
}
func (p *reader) layoutValue() Literal {
	s := p.t().span
	if p.accept("(") {
		values := []float64{p.take("NUMBER").value.(float64)}
		for p.accept(",") {
			values = append(values, p.take("NUMBER").value.(float64))
			if len(values) > 4 {
				fail("syntax", "Padding has at most four values", p.t().span)
			}
		}
		p.take(")")
		return Literal{"tuple", values, p.span(s)}
	}
	if p.t().kind == "ID" && p.t().value != "true" && p.t().value != "false" && p.t().value != "null" {
		return Literal{"enum", p.take("ID").value, s}
	}
	v := p.literal()
	if v.Kind == "number" && p.t().value == "fr" {
		p.take("ID")
		v.Kind = "fr"
		v.Span = p.span(s)
	}
	return v
}
func (p *reader) layout() []LayoutRule {
	p.take("{")
	out := []LayoutRule{}
	if p.t().kind != "}" {
		for {
			s := p.t().span
			name := ""
			var v Literal
			if p.t().kind == "ID" && p.ahead("=") {
				name = p.take("ID").value.(string)
				p.take("=")
				v = p.layoutValue()
			} else if p.t().kind == "NUMBER" && p.ahead(":") {
				x := p.take("NUMBER").value.(float64)
				p.take(":")
				y := p.take("NUMBER").value.(float64)
				name = "ratio"
				v = Literal{"ratio", []float64{x, y}, p.span(s)}
			} else {
				if p.t().kind == "," || p.t().kind == "}" || p.t().kind == "EOF" {
					fail("syntax", "Expected formatting rule", p.t().span)
				}
				for p.t().kind != "," && p.t().kind != "}" && p.t().kind != "EOF" {
					p.index++
				}
				name = "arrow"
				v = Literal{"arrow", strings.Join(strings.Fields(p.source[s.Start:p.span(s).End]), ""), p.span(s)}
			}
			out = append(out, LayoutRule{name, v, p.source[s.Start:p.span(s).End], p.span(s)})
			if len(out) > 32 {
				fail("layout-limit", "More than 32 rules", p.t().span)
			}
			if !p.accept(",") {
				break
			}
		}
	}
	p.take("}")
	return out
}
